mam/src/mam.rs

```diff
@@ -39,116 +37,148 @@ where
+            end
         };
-
-        let mut siblings = vec![0 as Trit; merkle::siblings_count(addresses.len()) * HASH_LENGTH];
-        merkle::siblings(&addresses, index, &mut siblings, curl1);
-        curl1.reset();
-        let root = merkle::root(&addresses[index], &siblings, index, curl1);
-        curl1.reset();
-        (siblings, root)
-    };
-
-    let next = {
-        let mut digest = [0 as Trit; iss::DIGEST_LENGTH];
-        let next_addrs: Vec<Vec<Trit>> = {
-            let mut addr = Vec::new();
-            for i in next_start..(next_start + next_count) {
-                merkle::key(seed, i, security, &mut key, curl1);
-                curl1.reset();
-
-                iss::digest_key::<Trit, C>(&key, &mut digest, curl1, curl2);
-                curl1.reset();
-                curl2.reset();
-                iss::address::<Trit, C>(&mut digest, curl1);
-                curl1.reset();
-                addr.push(digest[..iss::ADDRESS_LENGTH].to_vec());
-            }
-            addr
+        cursor = {
+            // get the hamming nonce, append and mask it to payload
+            curl_bak.clone_from_slice(&sponge.state());
+            H::search(security, 0, HASH_LENGTH / 3, sponge, bcurl).unwrap();
+            payload.extend_from_slice(&sponge.rate()[..HASH_LENGTH / 3]);
+            let end = payload.len();
+            sponge.state_mut().clone_from_slice(&curl_bak);
+            mask_slice(&mut payload[cursor..end], sponge);
+            bcurl.reset();
+            end
         };
-        curl1.reset();
-        curl2.reset();
-
-        let mut siblings = vec![0 as Trit; merkle::siblings_count(next_addrs.len()) * HASH_LENGTH];
-        merkle::siblings(&next_addrs, 0, &mut siblings, curl1);
-        curl1.reset();
-        merkle::root(&next_addrs[0], &siblings, 0, curl1)
-    };
-
-    curl1.reset();
-    curl2.reset();
-
-    merkle::key(seed, start + index, security, &mut key, curl1);
-    curl1.reset();
-    let mut payload = sign::<C, CB, H>(message, &next, &key, &siblings, security, curl1, bcurl);
-
-    {
-        // generate the message key for a root and an index.
-        let mut channel_key: [Trit; HASH_LENGTH] = [0; HASH_LENGTH];
-        // It copies `root` to `channel_key`
-        channel_key.clone_from_slice(&root[..HASH_LENGTH]);
-        // add `index` to it.
-        add_assign(&mut channel_key, index as isize);
-        mask::<C>(&mut payload, &channel_key, curl2);
+        curl_bak.clone_from_slice(&sponge.state());
+        sponge.reset();
+        iss::signature(&curl_bak[..HASH_LENGTH], &mut key, sponge);
+        sponge.state_mut().clone_from_slice(&curl_bak);
+        payload.extend_from_slice(&key);
+        payload.extend_from_slice(&pascal::encode(siblings.len() / HASH_LENGTH));
+        payload.extend_from_slice(&siblings);
+        let end = payload.len();
+        mask_slice(&mut payload[cursor..end], sponge);
+        sponge.reset();
```

Comment:
We should memset `curl_bak` as well so we don't leave state in memory. Just do another 
`curl_bak.clone_from_slice(&sponge.state())` here (as the latter has been reset to 0)